Flaky test: test_merge_tree_s3_failover/test.py::test_write_failover[0-15-2]

_Important: This issue was automatically generated and is used by CI for matching failures. DO NOT modify the body content. DO NOT remove labels._

Test name: test_merge_tree_s3_failover/test.py::test_write_failover[0-15-2]
Failure reason: AssertionError: Insert query should be failed, request 7
CI report: [Integration tests (arm_binary, distributed plan, 4/4)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=96454&sha=1dbff55dd5ad2ce8947db3ae3b60c16577b91a01&name_0=PR&name_1=Integration%20tests%20%28arm_binary%2C%20distributed%20plan%2C%204%2F4%29)

Failing test history: [cidb](https://play.clickhouse.com#V0lUSAogICAgOTAgQVMgaW50ZXJ2YWxfZGF5cwpTRUxFQ1QKICAgIHRvU3RhcnRPZkRheShjaGVja19zdGFydF90aW1lKSBBUyBkYXksCiAgICBjb3VudCgpIEFTIGZhaWx1cmVzLAogICAgZ3JvdXBVbmlxQXJyYXkocHVsbF9yZXF1ZXN0X251bWJlcikgQVMgcHJzLAogICAgYW55KHJlcG9ydF91cmwpIEFTIHJlcG9ydF91cmwKRlJPTSBjaGVja3MKV0hFUkUgKG5vdygpIC0gdG9JbnRlcnZhbERheShpbnRlcnZhbF9kYXlzKSkgPD0gY2hlY2tfc3RhcnRfdGltZQogICAgQU5EIHRlc3RfbmFtZSA9ICd0ZXN0X21lcmdlX3RyZWVfczNfZmFpbG92ZXIvdGVzdC5weTo6dGVzdF93cml0ZV9mYWlsb3ZlclswLTE1LTJdJwogICAgLS0gQU5EIGNoZWNrX25hbWUgPSAnSW50ZWdyYXRpb24gdGVzdHMgKGFybV9iaW5hcnksIGRpc3RyaWJ1dGVkIHBsYW4sIDQvNCknCiAgICBBTkQgdGVzdF9zdGF0dXMgSU4gKCdGQUlMJywgJ0VSUk9SJykKICAgIEFORCAocHVsbF9yZXF1ZXN0X251bWJlciA9IDAgT1IgYmFzZV9yZWYgSU4gKCdtYXN0ZXInKSkKICAgIEFORCB0ZXN0X2NvbnRleHRfcmF3IExJS0UgJyVBc3NlcnRpb25FcnJvcjogSW5zZXJ0IHF1ZXJ5IHNob3VsZCBiZSBmYWlsZWQsIHJlcXVlc3QgNyUnCkdST1VQIEJZIGRheQpPUkRFUiBCWSBkYXkgREVTQwo=)

Test output:
```
File: test_merge_tree_s3_failover/test.py:145 - in test_write_failover
    assert positive, "Insert query should be failed, request {}".format(request)
E   AssertionError: Insert query should be failed, request 7
E   assert False
```

--- a/tests/integration/test_merge_tree_s3_failover/s3_endpoint/endpoint.py
+++ b/tests/integration/test_merge_tree_s3_failover/s3_endpoint/endpoint.py
@@ -11,12 +11,15 @@ mutex = Lock()
 
 
 @route("/fail_request/<_request_number>")
-def fail_request(_request_number):
+@route("/fail_request/<_request_number>/<_method>")
+def fail_request(_request_number, _method=None):
     request_number = int(_request_number)
     if request_number > 0:
         cache["request_number"] = request_number
+        cache["fail_method"] = _method.upper() if _method else None
     else:
         cache.pop("request_number", None)
+        cache.pop("fail_method", None)
     return "OK"
 
 
@@ -50,11 +53,15 @@ def server(_bucket, _path):
 
     mutex.acquire()
     try:
-        if cache.get("request_number", None):
+        fail_method = cache.get("fail_method", None)
+        if cache.get("request_number", None) and (
+            fail_method is None or request.method == fail_method
+        ):
             request_number = cache.pop("request_number") - 1
             if request_number > 0:
                 cache["request_number"] = request_number
             else:
+                cache.pop("fail_method", None)
                 response.status = 500
                 response.content_type = "text/xml"
                 return '<?xml version="1.0" encoding="UTF-8"?><Error><Code>ExpectedError</Code><Message>Expected Error</Message><RequestId>txfbd566d03042474888193-00608d7537</RequestId></Error>'

--- a/tests/integration/test_merge_tree_s3_failover/test.py
+++ b/tests/integration/test_merge_tree_s3_failover/test.py
@@ -41,10 +41,13 @@ def run_endpoint(cluster):
     logging.info("S3 endpoint started")
 
 
-def fail_request(cluster, request):
+def fail_request(cluster, request, method=None):
+    url = "http://resolver:8080/fail_request/{}".format(request)
+    if method:
+        url += "/{}".format(method)
     response = cluster.exec_in_container(
         cluster.get_container_id("resolver"),
-        ["curl", "-s", "http://resolver:8080/fail_request/{}".format(request)],
+        ["curl", "-s", url],
     )
     assert response == "OK", 'Expected "OK", but got "{}"'.format(response)
 
@@ -133,8 +136,11 @@ def test_write_failover(
     success_count = 0
 
     for request in range(request_count + debug_request_count + 1):
-        # Fail N-th request to S3.
-        fail_request(cluster, request + 1)
+        # Fail N-th PUT request to S3.
+        # Use PUT method filter to avoid counting background GET/DELETE
+        # requests (e.g. cleanup of failed parts) that could consume the
+        # fail counter and make the test flaky.
+        fail_request(cluster, request + 1, "PUT")
 
         data = "('2020-03-01',0,'data'),('2020-03-01',1,'data')"
         try:

